src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```diff
@@ -0,0 +1,135 @@
+// Also DOM is only available on the main (browser) thread.
+// The calls to InvokeAsync methods are dispatched synchronously and the returned Task is only resolved after the main thread finished the callback.
+internal sealed class WebAssemblyDispatcher : Dispatcher
+{
+    public static readonly Dispatcher Instance = new WebAssemblyDispatcher();
+    private readonly SynchronizationContext? _context;
+
+    private WebAssemblyDispatcher()
+    {
+        // capture the JSSynchronizationContext from the main thread.
+        _context = SynchronizationContext.Current;
+    }
+
+    public override bool CheckAccess() => SynchronizationContext.Current == _context || _context == null;
+
+    public override Task InvokeAsync(Action workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            workItem();
+            return Task.CompletedTask;
+        }
+
+        _context!.InvokeAsync(workItem);
+        return Task.CompletedTask;
+    }
+
+    public override Task InvokeAsync(Func<Task> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            return workItem();
+        }
+
+        return _context!.InvokeAsync(workItem);
+    }
+
+    public override Task<TResult> InvokeAsync<TResult>(Func<TResult> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            return Task.FromResult(workItem());
+        }
+
+        return _context!.InvokeAsync(static (workItem) => Task.FromResult(workItem()), workItem);
+    }
+
+    public override Task<TResult> InvokeAsync<TResult>(Func<Task<TResult>> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            return workItem();
+        }
+
+        return _context!.InvokeAsync(workItem);
+    }
+}
+
+internal static class SynchronizationContextExtension
+{
+    public static void InvokeAsync(this SynchronizationContext self, Action body)
+    {
+        Exception? exc = default;
+        self.Send((_) =>
+        {
+            try
+            {
+                body();
+            }
+            catch (Exception ex)
+            {
+                exc = ex;
+            }
+        }, null);
+        if (exc != null)
+        {
+            throw exc;
+        }
+    }
+
+    public static TRes InvokeAsync<TRes>(this SynchronizationContext self, Func<TRes> body)
+    {
+        TRes? value = default;
+        Exception? exc = default;
+        self.Send((_) =>
+        {
+            try
+            {
+                value = body();
+            }
+            catch (Exception ex)
+            {
+                exc = ex;
+            }
+        }, null);
+        if (exc != null)
+        {
+            throw exc;
+        }
+        return value!;
+    }
```

Comment:
The logic around async and exception handling looks strange to me. At least, it will behave very differently from how `RendererSynchronizationContext` does.

* `Dispatcher.InvokeAsync(Action)`
   * With `WebAssemblyDispatcher`, it blocks the caller until the action is fully complete. If the action throws, then it re-throws instead of returning any task.
   * With `RendererSynchronizationContextDispatcher`, it doesn't block the caller, and returns a `Task` that completes when the action is complete. If the action throws, the `Task` is faulted.
* `Dispatcher.InvokeAsync(Func<Task>)`
  * With `WebAssemblyDispatcher`, it blocks the caller while the dispatch is in process (up until the recipient yields), and then returns the `Task` returned by the recipient. If the recipient throws synchronously, then it re-throws instead of returning any task.
   * With `RendererSynchronizationContextDispatcher`, it does not block the caller, and returns a a `Task` that completes when the dispatch *and* recipient's Task is complete. If the recipent throws synchronously, the returned `Task` is faulted.

Are these differences in behavior intentional? My guess is we should behave the same as `RendererSynchronizationContextDispatcher`. Of course, the ideal would be to actually use `RendererSynchronizationContextDispatcher` as-is without reimplementing it.

Reply:
Similarly please see how `RendererSynchronizationContext` deals with unhandled exceptions by calling `OnUnhandledException` on the `Dispatcher`.